src/v/compression/lz4_decompression_buffers.cc

```diff
@@ -0,0 +1,228 @@
+          .state = alloc_ctx::allocation_state::no_buffers_allocated,
+        };
+    }
+}
+
+bool lz4_decompression_buffers::has_reserved_buffers() const {
+    return _reservation.current() == 0;
+}
+
+ss::future<std::optional<ssx::semaphore_units>>
+lz4_decompression_buffers::reserve_buffers() {
+    if (_disabled) {
+        co_return std::nullopt;
+    }
+
+    co_return co_await ss::get_units(_reservation, 1);
+}
+
+bool lz4_decompression_buffers::alloc_ctx::is_managed_address(
+  const void* const address) const {
+    return address == input_buffer.get() || address == output_buffer.get();
+}
+
+lz4_decompression_buffers::alloc_ctx& lz4_decompression_buffers::buffers() {
+    vassert(
+      has_reserved_buffers(), "attempt to access buffers when none reserved");
+    return _buffers;
+}
+
+size_t lz4_decompression_buffers::min_alloc_threshold() const {
+    return _min_alloc_threshold;
+}
+
+LZ4F_CustomMem lz4_decompression_buffers::custom_mem_alloc() {
+    // If custom allocation is disabled, setting all alloc functions to null
+    // makes lz4 fall back to malloc, calloc and free.
+    if (_disabled) {
+        return {
+          .customAlloc = nullptr,
+          .customCalloc = nullptr,
+          .customFree = nullptr,
+          .opaqueState = nullptr};
+    }
+
+    return {
+      .customAlloc = alloc_lz4_obj,
+      .customCalloc = nullptr,
+      .customFree = free_lz4_obj,
+      .opaqueState = this};
+}
+
+static thread_local std::unique_ptr<lz4_decompression_buffers>
+  _buffers_instance;
+
+static constexpr auto default_buf_size = 4_MiB + 128_KiB;
+static constexpr auto default_min_alloc_threshold = 4_MiB;
+
+void init_lz4_decompression_buffers(
+  size_t buffer_size, size_t min_alloc_threshold, bool prealloc_disabled) {
+    if (!_buffers_instance) {
+        _buffers_instance = std::make_unique<lz4_decompression_buffers>(
+          buffer_size, min_alloc_threshold, prealloc_disabled);
+    }
+}
+
+void reset_lz4_decompression_buffers() {
+    if (_buffers_instance) {
+        _buffers_instance.reset();
+    }
+}
+
+lz4_decompression_buffers& lz4_decompression_buffers_instance() {
+    if (unlikely(!_buffers_instance)) {
+        init_lz4_decompression_buffers(
+          default_buf_size, default_min_alloc_threshold);
+    }
+
+    return *_buffers_instance;
+}
+
+} // namespace compression
+
+// During a typical lz4 decompression operation the following LZ4F_malloc calls
+// will be processed via this alloc function:
+// 1. Allocation for the tmp input buffer: this can be a maximum of 4MiB + 4
+// bytes
+// 2. Allocation for the tmp output buffer: this can be a maximum of 4MiB +
+// 128KiB
+// These two calls will typically happen once per decompression context, and are
+// preceded by calls to LZ4F_free to first free up the two buffers.
+void* alloc_lz4_obj(void* state, size_t size) {
+    auto* st = static_cast<compression::lz4_decompression_buffers*>(state);
```

Comment:
How certain are we that `size` will be under 4MiB+128KiB? Should we check `size` and fail in some reasonable way (throwing an exception I guess would be the way, but I'm not sure what that means for the state transitions, so maybe even crashing makes sense). I get that that's why we're making this configurable, but:
1. a user can change it to whatever they want
2. this seems like a very hard assumption we're putting on lz4 -- if it changes, or if there's a bug or somesuch, it seems like we could get into some trouble

Reply:
I think it makes sense to crash in case of a larger size. Right now the upper size limit is guaranteed, but lz4 can change in future, and then we might get a larger block size.

This brings up another point, do we want to expose the buffer size as a config parameter? The reason I added this was for cases where the user did not want the overhead of losing 8MiB when they perhaps know there is no need for it (maybe they do not use compression or lz4) but now we have a setting to turn off the allocation, so perhaps the size should not be exposed to the user.

Reply:
>do we want to expose the buffer size as a config parameter

Right, I would lean towards not exposing it, other than enabling/disabling the extra buffers. Otherwise it seems too easy to get Redpanda to crash

Reply:
removed the two size related settings